Features/ManageabilityPkg/Universal/IpmiSmbiosTransferDxe/IpmiSmbiosTransferDxe.c

```diff
@@ -0,0 +1,293 @@
+/** @file
+
+  A driver that sends SMBIOS tables to an OpenBMC receiver
+
+  SPDX-FileCopyrightText: copyright (c) 2022-2024, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+
+  SPDX-License-Identifier: BSD-2-Clause-Patent
+
+**/
+#include <Guid/SmBios.h>
+
+#include <IndustryStandard/SmBios.h>
+
+#include <Library/BaseMemoryLib.h>
+#include <Library/DebugLib.h>
+#include <Library/UefiBootServicesTableLib.h>
+#include <Library/UefiRuntimeServicesTableLib.h>
+#include <Library/UefiLib.h>
+#include <Library/MemoryAllocationLib.h>
+#include <Library/ReportStatusCodeLib.h>
+#include <Library/BaseCryptLib.h>
+
+#include <Protocol/IpmiBlobTransfer.h>
+
+#define SMBIOS_HASH_VARIABLE                   L"SmbiosHash"
+#define SMBIOS_TRANSFER_DEBUG                  DEBUG_MANAGEABILITY
+#define SMBIOS_EC_DESC_NO_SMBIOS_TABLE         "No SMBIOS table installed"
+#define SMBIOS_EC_DESC_SMBIOS_TRANSFER_FAILED  "Failed to send SMBIOS tables to BMC"
+
+/**
+  This function will calculate smbios hash, compares it with stored
+  hash, updates UEFI variable if smbios data changed and return status.
+
+  @param[in]    The smbios data to detect if changed.
+  @param[in]    The smbios data size.
+  @retval TRUE  If smbios data changed, otherwise FALSE.
+**/
+BOOLEAN
+DetectSmbiosChange (
+  UINT8   *SmbiosData,
+  UINT32  SmbiosDataSize
+  )
+{
+  BOOLEAN     UpdateRequired = TRUE;
+  BOOLEAN     DeleteVar      = FALSE;
+  BOOLEAN     Response;
+  UINT8       ComputedHashValue[SHA256_DIGEST_SIZE];
+  UINT8       StoredHashValue[SHA256_DIGEST_SIZE];
+  UINTN       StoredHashValueSize;
+  EFI_STATUS  Status;
+
+  Response = Sha256HashAll ((VOID *)SmbiosData, SmbiosDataSize, ComputedHashValue);
+
+  if (Response == FALSE) {
+    // Delete the SmbiosHash variable and continue with smbios transfer to BMC
+    Status = gRT->SetVariable (
+                    SMBIOS_HASH_VARIABLE,
+                    &gManageabilityVariableGuid,
+                    EFI_VARIABLE_NON_VOLATILE | EFI_VARIABLE_BOOTSERVICE_ACCESS,
+                    0,
+                    NULL
+                    );
+    if (EFI_ERROR (Status)) {
+      DEBUG ((DEBUG_ERROR, "%a: Failed to delete UEFI Variable SmbiosHash %r\n", __func__, Status));

```

Comment:
The variable keeps the hash value of smbios table. We compare it to know if the smbios table has been changed or not between this boot and previous boot. If for some reason we cannot get the hash value, we delete the variable because its value is not referring to the smbios table in this boot anymore. On next boot, BIOS will send smbios table to BMC because there is no variable.